src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs

```diff
@@ -240,44 +248,120 @@ private Task WriteMultiEndpointMessageAsync(ServiceMessage serviceMessage, Func<
             }
 
             return Task.WhenAll(routed);
-        }
-
-        private void OnAdd(HubServiceEndpoint endpoint)
-        {
-            if (!endpoint.Hub.Equals(_hubName, StringComparison.OrdinalIgnoreCase))
-            {
-                return;
-            }
-            _ = AddHubServiceEndpointAsync(endpoint);
-        }
-
-        private Task AddHubServiceEndpointAsync(HubServiceEndpoint endpoint)
-        {
-            // TODO: create container and trigger server ping.
-
-            // do tasks when !endpoint.ScaleTask.IsCanceled or local timeout check not finish
-            endpoint.CompleteScale();
-            return Task.CompletedTask;
-        }
-
-        private void OnRemove(HubServiceEndpoint endpoint)
-        {
-            if (!endpoint.Hub.Equals(_hubName, StringComparison.OrdinalIgnoreCase))
-            {
-                return;
-            }
-            _ = RemoveHubServiceEndpointAsync(endpoint);
-        }
-
-        private Task RemoveHubServiceEndpointAsync(HubServiceEndpoint endpoint)
-        {
-            // TODO: trigger offline ping and wait to remove container.
-
-            // finally set task complete when timeout
-            endpoint.CompleteScale();
-            return Task.CompletedTask;
-        }
-
+        }
+
+        private void OnAdd(HubServiceEndpoint endpoint)
+        {
+            if (!endpoint.Hub.Equals(_hubName, StringComparison.OrdinalIgnoreCase))
+            {
+                return;
+            }
+            _ = AddHubServiceEndpointAsync(endpoint);
+        }
+
+        private async Task AddHubServiceEndpointAsync(HubServiceEndpoint endpoint)
+        {
+            var container = _generator(endpoint);
+            endpoint.ConnectionContainer = container;
+
+            try
+            {
+                _ = container.StartAsync();
```

Comment:
do we need to await container until it successfully started?

Reply:
not required. it has multiple connections inside and the procedure might be longer then expected. message send during the time would be no harm cause no client is listening.

---
In reply to: [392102714](https://github.com/Azure/azure-signalr/pull/841#discussion_r392102714) [](ancestors = 392102714)

Reply:
if the server connection is not yet established, will the message send throw ConnectionNotActiveException?